state-chain/pallets/cf-threshold-signature/src/tests.rs

```diff
@@ -732,3 +748,1179 @@ mod failure_reporting {
+		RuntimeOrigin::signed(CHARLIE),
+		keygen_ceremony_id,
+		Ok(NEW_AGG_PUB_KEY_PRE_HANDOVER)
+	));
+
+	// This time we should have enough votes for consensus.
+	assert!(KeygenResolutionPendingSince::<Test, _>::exists());
+	assert_eq!(<EthereumThresholdSigner as VaultRotator>::status(), AsyncResult::Pending);
+	if let VaultRotationStatus::AwaitingKeygen {
+		ceremony_id: keygen_ceremony_id_from_status,
+		response_status,
+		keygen_participants,
+		new_epoch_index,
+	} = PendingVaultRotation::<Test, _>::get().unwrap()
+	{
+		assert_eq!(keygen_ceremony_id, keygen_ceremony_id_from_status);
+		assert_eq!(
+			response_status
+				.success_votes()
+				.get(&NEW_AGG_PUB_KEY_PRE_HANDOVER)
+				.expect("new key should have votes"),
+			&(ALL_CANDIDATES.len() as AuthorityCount)
+		);
+		assert_eq!(keygen_participants, BTreeSet::from_iter(ALL_CANDIDATES.iter().cloned()));
+		assert_eq!(new_epoch_index, rotation_epoch);
+	} else {
+		panic!("Expected to be in AwaitingKeygen state");
+	}
+	<EthereumThresholdSigner as Hooks<BlockNumberFor<Test>>>::on_initialize(1);
+
+	assert!(matches!(
+		PendingVaultRotation::<Test, _>::get().unwrap(),
+		VaultRotationStatus::AwaitingKeygenVerification { .. }
+	));
+
+	let cfes = [ALICE]
+		.iter()
+		.map(|id| MockCfe { id: *id, behaviour: CfeBehaviour::Success })
+		.collect::<Vec<_>>();
+	run_cfes_on_sc_events(&cfes);
+
+	assert_eq!(
+		<EthereumThresholdSigner as VaultRotator>::status(),
+		AsyncResult::Ready(VaultRotationStatusOuter::KeygenComplete)
+	);
+
+	assert!(matches!(
+		PendingVaultRotation::<Test, _>::get().unwrap(),
+		VaultRotationStatus::KeygenVerificationComplete { .. }
+	));
+
+	const SHARING_PARTICIPANTS: [u64; 2] = [ALICE, BOB];
+	EthereumThresholdSigner::key_handover(
+		BTreeSet::from(SHARING_PARTICIPANTS),
+		BTreeSet::from_iter(ALL_CANDIDATES.iter().copied()),
+		rotation_epoch,
+	);
+	assert_eq!(<EthereumThresholdSigner as VaultRotator>::status(), AsyncResult::Pending);
+
+	let handover_ceremony_id = current_ceremony_id();
+	for p in ALL_CANDIDATES {
+		assert_ok!(EthereumThresholdSigner::report_key_handover_outcome(
+			RuntimeOrigin::signed(*p),
+			handover_ceremony_id,
+			Ok(NEW_AGG_PUB_KEY_POST_HANDOVER)
+		));
+	}
+
+	<EthereumThresholdSigner as Hooks<BlockNumberFor<Test>>>::on_initialize(1);
+	assert_eq!(<EthereumThresholdSigner as VaultRotator>::status(), AsyncResult::Pending);
+
+	//assert_last_event!(crate::Event::ThresholdSignatureRequest { .. });
+
+	//assert_last_event!(crate::Event::KeyHandoverSuccess { .. });
+
+	assert!(matches!(
+		PendingVaultRotation::<Test, _>::get().unwrap(),
+		VaultRotationStatus::AwaitingKeyHandoverVerification { .. }
+	));
+
+	run_cfes_on_sc_events(&cfes);
+
+	assert_eq!(
+		<EthereumThresholdSigner as VaultRotator>::status(),
+		AsyncResult::Ready(VaultRotationStatusOuter::KeyHandoverComplete)
+	);
+
+	//println!("{:?}", System::events());
+	//assert_last_event!(crate::Event::KeyHandoverVerificationSuccess { .. });
+
```

Comment:
Guess we we want to uncomment these lines

Reply:
There asserts didnt work anymore since other events are also emitted now and the the actual event we want to assert is not the last but second last (sometimes third last). In the last commit I updated the ```assert_last_event``` macro to assert arbitrary number of last events and then updated these asserts.